ai/cohere/v0/client.go

```diff
@@ -0,0 +1,76 @@
+package cohere
+
+import (
+	"context"
+	"sync"
+
+	cohereSDK "github.com/cohere-ai/cohere-go/v2"
+	cohereClientSDK "github.com/cohere-ai/cohere-go/v2/client"
+	"github.com/cohere-ai/cohere-go/v2/core"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+type cohereClient struct {
+	sdkClient cohereClientInterface
+	logger    *zap.Logger
+	lock      sync.Mutex
+}
+
+type cohereClientInterface interface {
+	Chat(ctx context.Context, request *cohereSDK.ChatRequest, opts ...core.RequestOption) (*cohereSDK.NonStreamedChatResponse, error)
+	Embed(ctx context.Context, request *cohereSDK.EmbedRequest, opts ...core.RequestOption) (*cohereSDK.EmbedResponse, error)
+	Rerank(ctx context.Context, request *cohereSDK.RerankRequest, opts ...core.RequestOption) (*cohereSDK.RerankResponse, error)
+}
+
+func newClient(apiKey string, logger *zap.Logger) *cohereClient {
+	client := cohereClientSDK.NewClient(cohereClientSDK.WithToken(apiKey))
+	return &cohereClient{sdkClient: client, logger: logger, lock: sync.Mutex{}}
+}
+
+func (cl *cohereClient) generateEmbedding(request cohereSDK.EmbedRequest) (cohereSDK.EmbedResponse, error) {
+	respPtr, err := cl.sdkClient.Embed(
+		context.TODO(),
+		&request,
+	)
+	if err != nil {
+		panic(err)
+	}
+	resp := cohereSDK.EmbedResponse{
+		EmbeddingsFloats: respPtr.EmbeddingsFloats,
```

Comment:
Important

Will need to get the input & output tokens.

Reply:
Fixed